plugwise/helper.py

```diff
@@ -175,103 +211,112 @@ async def request(
         if len(self._locations) == 0 and self._smile_legacy:
-            appliances = set()
-            home_location = 0
-
-            # Add Anna appliances
-            for appliance in self._appliances:
-                appliances.add(appliance.attrib["id"])
-
-            if self.smile_type == "thermostat":
-                self._loc_data[0] = {
-                    "name": "Legacy Anna",
-                    "types": {"temperature"},
-                    "members": appliances,
-                }
-            if self.smile_type == "stretch":
-                self._loc_data[0] = {
-                    "name": "Legacy Stretch",
-                    "types": {"power"},
-                    "members": appliances,
-                }
-
-            self._home_location = home_location
-
+            self.locations_legacy()
             return
 
         for location in self._locations:
-            location_name = location.find("name").text
-            location_id = location.attrib["id"]
-            location_types = set()
-            location_members = set()
+            loc.name = location.find("name").text
+            loc.id = location.attrib["id"]
+            loc.types = set()
+            loc.members = set()
 
             # Group of appliances
             locator = ".//appliances/appliance"
             if location.find(locator) is not None:
                 for member in location.findall(locator):
-                    location_members.add(member.attrib["id"])
-
-            if location_name == "Home":
-                home_location = location_id
-                location_types.add("home")
+                    loc.members.add(member.attrib["id"])
 
-                for location_type in types_finder(location):
-                    location_types.add(location_type)
+            # Specials
+            loc = self.locations_specials(loc, location)
 
-            # Legacy P1 right location has 'services' filled
-            # test data has 5 for example
-            locator = ".//services"
-            if (
-                self._smile_legacy
-                and self.smile_type == "power"
-                and len(location.find(locator)) > 0
-            ):
-                # Override location name found to match
-                location_name = "Home"
-                home_location = location_id
-                location_types.add("home")
-                location_types.add("power")
-
-            self._loc_data[location_id] = {
-                "name": location_name,
-                "types": location_types,
-                "members": location_members,
+            self._loc_data[loc.id] = {
+                "name": loc.name,
+                "types": loc.types,
+                "members": loc.members,
             }
 
-        self._home_location = home_location
-
         return
 
+    def appliance_stretch_class_finder(self, appliance, appl):
```

Comment:
The function-name is too specific, the 1st part is covering the devices connected to a Stretch, at the moment we are only supporting Circles and Stealths, the 2nd part is covering Plugs connected to an Adam.

I would suggest 'energy_device_info_finder' or something similar.

Reply:
Good call!